common/unified/distributed/partition_kernels.cpp

```diff
@@ -0,0 +1,174 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/distributed/partition_kernels.hpp"
+
+
+#include "common/unified/base/kernel_launch.hpp"
+#include "common/unified/base/kernel_launch_reduction.hpp"
+#include "core/components/prefix_sum.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace GKO_DEVICE_NAMESPACE {
+namespace partition {
+
+
+using distributed::comm_index_type;
+
+void count_ranges(std::shared_ptr<const DefaultExecutor> exec,
+                  const Array<comm_index_type>& mapping, size_type& num_ranges)
+{
+    Array<size_type> result{exec, 1};
+    run_kernel_reduction(
+        exec,
+        [] GKO_KERNEL(auto i, auto mapping) {
+            auto cur_part = mapping[i];
+            auto prev_part = i == 0 ? comm_index_type{-1} : mapping[i - 1];
+            return cur_part != prev_part ? 1 : 0;
+        },
+        [] GKO_KERNEL(auto a, auto b) { return a + b; },
+        [] GKO_KERNEL(auto a) { return a; }, size_type{}, result.get_data(),
+        mapping.get_num_elems(), mapping);
+    num_ranges = exec->copy_val_to_host(result.get_const_data());
+}
+
+
+template <typename GlobalIndexType>
+void build_from_contiguous(std::shared_ptr<const DefaultExecutor> exec,
+                           const Array<GlobalIndexType>& ranges,
+                           GlobalIndexType* range_bounds,
+                           comm_index_type* part_ids)
+{
+    run_kernel(
+        exec,
+        [] GKO_KERNEL(auto i, auto ranges, auto bounds, auto ids) {
+            bounds[i + 1] = ranges[i + 1];
+            ids[i] = i;
+        },
```

Comment:
range_bounds[0]?

Reply:
I think that's taken care of because it's initialized to 0 in the constructor. But, for robustness w.r.t. future changes in the constructor, it might be better to explicitly set it here.

Reply:
I set it now explicitly in the kernel.